prow/cmd/hmac/main.go

```diff
@@ -0,0 +1,389 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"crypto/rand"
+	"encoding/hex"
+	"errors"
+	"flag"
+	"fmt"
+	"os"
+	"sort"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"sigs.k8s.io/yaml"
+
+	"k8s.io/test-infra/pkg/flagutil"
+	"k8s.io/test-infra/prow/cmd/hmac/updater"
+	"k8s.io/test-infra/prow/config"
+	prowflagutil "k8s.io/test-infra/prow/flagutil"
+	"k8s.io/test-infra/prow/github"
+	"k8s.io/test-infra/prow/logrusutil"
+)
+
+type options struct {
+	configPath string
+
+	dryRun     bool
+	github     prowflagutil.GitHubOptions
+	kubernetes prowflagutil.KubernetesOptions
+
+	hookUrl                  string
+	hmacTokenSecretNamespace string
+	hmacTokenSecretName      string
+	hmacTokenKey             string
+}
+
+func (o *options) validate() error {
+	for _, group := range []flagutil.OptionGroup{&o.kubernetes, &o.github} {
+		if err := group.Validate(o.dryRun); err != nil {
+			return err
+		}
+	}
+
+	if o.configPath == "" {
+		return errors.New("required flag --config-path was unset")
+	}
+	if o.hookUrl == "" {
+		return errors.New("required flag --hook-url was unset")
+	}
+	if o.hmacTokenSecretName == "" {
+		return errors.New("required flag --hmac-token-secret-name was unset")
+	}
+	if o.hmacTokenKey == "" {
+		return errors.New("required flag --hmac-token-key was unset")
+	}
+
+	return nil
+}
+
+func gatherOptions(fs *flag.FlagSet, args ...string) options {
+	var o options
+
+	o.github.AddFlags(fs)
+	o.kubernetes.AddFlags(fs)
+
+	fs.StringVar(&o.configPath, "config-path", "", "Path to config.yaml.")
+	fs.BoolVar(&o.dryRun, "dry-run", true, "Dry run for testing. Uses API tokens but does not mutate.")
+
+	fs.StringVar(&o.hookUrl, "hook-url", "", "Prow hook external webhook URL (e.g. https://prow.k8s.io/hook).")
+	fs.StringVar(&o.hmacTokenSecretNamespace, "hmac-token-secret-namespace", "default", "Name of the namespace on the cluster where the hmac-token secret is in.")
```

Comment:
This should just use whatever namespace is specified by the context that we select.

In production we should limit this to the current namespace, IE this should have a role, not a clusterrole

Reply:
Could you provide more context on this?

Based on the current design, this tool will be run by a postsubmit Prow job, so it will be a pod that runs in the `prow-trusted` cluster, so there won't be a `Role` associated with it.